composer/trainer/trainer.py

```diff
@@ -1106,98 +1182,108 @@ def _train_batch(self, microbatches: Sequence[Batch], ddp_sync: bool = True):
+            self.engine.run_event(Event.AFTER_FORWARD)
 
-                if self.deepspeed_enabled:
-                    self.state.deepspeed_model.backward(self.state.loss)
+            # loss
+            self.engine.run_event(Event.BEFORE_LOSS)
 
-                    # This is the same loss scaling and reporting we skipped earlier.
-                    for loss in ensure_tuple(self.state.loss):
-                        loss.mul_(self.state.batch_num_samples / current_batch_size)
-                        total_loss += loss.detach().clone()
-                else:
-                    for loss in ensure_tuple(self.state.loss):
-                        loss.backward(create_graph=self._backwards_create_graph)
+            with self.state.precision_context:
+                self.state.loss = self._original_model.loss(self.state.outputs, self.state.batch)
+
+            # We always want to scale loss by the grad_accum before the backwards pass and
+            # also for sake of metrics. Complicating matters, the DeepSpeed engine does its
+            # own scaling when we call `.backward`, but this isn't in place so we still need
+            # to scale for sake of metrics after the `.backward` call.
+
+            # Loss is added to losses with clone to not scale the loss for the step printout
+            # Likely need to look into the performance impact
+            if not self.deepspeed_enabled:
+                for loss in ensure_tuple(self.state.loss):
+                    loss.mul_(microbatch_num_samples / current_batch_size)
+                    total_loss += loss.detach().clone()
+
+            assert self.state.loss is not None
+            self.engine.run_event(Event.AFTER_LOSS)
+
+            # backward
+            self.engine.run_event(Event.BEFORE_BACKWARD)
 
-                self.engine.run_event(Event.AFTER_BACKWARD)
+            if use_grad_scaling:
+                self.state.loss = cast(torch.Tensor, self.state.scaler.scale(self.state.loss))
 
             if self.deepspeed_enabled:
-                self.state.deepspeed_model.step()
-
-        # Unscale gradients before `Event.AFTER_TRAIN_BATCH`
-        if use_grad_scaling:
-            for optimizer in ensure_tuple(self.state.optimizers):
-                self.state.scaler.unscale_(optimizer)
-
-        # clip gradients if the magnitude is too large
-        if not self.deepspeed_enabled and self._grad_clip_norm is not None:
-            torch.nn.utils.clip_grad_norm_(
-                parameters=self.state.model.parameters(),
-                max_norm=self._grad_clip_norm,
-            )
+                self.state.deepspeed_model.backward(self.state.loss)
+
+                # This is the same loss scaling and reporting we skipped earlier.
+                for loss in ensure_tuple(self.state.loss):
+                    loss.mul_(self.state.batch_num_samples / current_batch_size)
```

Comment:
is `batch_num_samples` still being updated on state? I see that old lines 1129 and 1130 were removed.
I'd be OK with making the num samples and num tokens local to only the trainer, in which case could you remove these variables from the state?

Reply:
It's still set each iteration and used in a few places. We could pass it around as parameters, but imo fine to leave in `state` -- still "feels" right to be part of state

Reply:
I must be blind, but which lines is it being updated on?

Reply:
Ah yes, I see on lines 1002 and 1003.